File: bootstrap/src/main/java/net/dontdrinkandroot/wicket/bootstrap/component/dropdown/DropDownMenu.java
/**
 * Copyright (C) 2012-2014 Philip W. Sorst <philip@sorst.net>
 * and individual contributors as indicated
 * by the @authors tag.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.dontdrinkandroot.wicket.bootstrap.component.dropdown;

import org.apache.wicket.AttributeModifier;
import org.apache.wicket.markup.ComponentTag;
import org.apache.wicket.markup.html.panel.Panel;
import org.apache.wicket.markup.repeater.RepeatingView;
import org.apache.wicket.model.Model;

import net.dontdrinkandroot.wicket.behavior.CssClassAppender;
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass;


public abstract class DropDownMenu extends Panel
{

	public DropDownMenu(String id)
	{

		super(id);
	}

	@Override
	protected void onInitialize()
	{

		super.onInitialize();

		this.add(new CssClassAppender(BootstrapCssClass.DROPDOWN_MENU));
		this.add(new AttributeModifier("role", Model.of("menu")));
		// TODO: include aria-labelledby

		RepeatingView itemView = new RepeatingView("item");
		this.populateItems(itemView);
		this.add(itemView);
	}

	@Override
	protected void onComponentTag(ComponentTag tag)
	{
		tag.setName("ul");
		super.onComponentTag(tag);
	}

	protected abstract void populateItems(RepeatingView itemView);
}

File: bootstrap/src/main/kotlin/net/dontdrinkandroot/wicket/bootstrap/component/dropdown/DropdownMenu.kt
package net.dontdrinkandroot.wicket.bootstrap.component.dropdown

import net.dontdrinkandroot.wicket.behavior.CssClassAppender
import net.dontdrinkandroot.wicket.bootstrap.component.item.ItemContainer
import net.dontdrinkandroot.wicket.bootstrap.component.item.ItemView
import net.dontdrinkandroot.wicket.bootstrap.css.BootstrapCssClass
import net.dontdrinkandroot.wicket.bootstrap.css.DropdownAlignment
import net.dontdrinkandroot.wicket.css.CssClass
import org.apache.wicket.AttributeModifier
import org.apache.wicket.markup.html.panel.Panel
import org.apache.wicket.model.IModel
import org.apache.wicket.model.Model

abstract class DropdownMenu(
    id: String,
    alignmentModel: IModel<DropdownAlignment?> = Model(null),
) : Panel(id), ItemContainer {

    init {
        this.add(CssClassAppender(BootstrapCssClass.DROPDOWN_MENU))
        this.add(CssClassAppender(alignmentModel))
        this.add(AttributeModifier("role", Model.of("menu")))
        // TODO: include aria-labelledby
        val itemView = ItemView("item")
        populateItems(itemView)
        this.add(itemView)
    }

    override val linkClass: CssClass?
        get() = BootstrapCssClass.DROPDOWN_ITEM

    abstract fun populateItems(itemView: ItemView)
}

fun dropdownMenu(
    id: String,
    alignmentModel: IModel<DropdownAlignment?> = Model(null),
    populateItemsHandler: ItemView.() -> Any?
) = object : DropdownMenu(id, alignmentModel) {
    override fun populateItems(itemView: ItemView) {
        populateItemsHandler(itemView)
    }
}